src/dmd/semantic3.d

```diff
@@ -1074,7 +1074,7 @@ private extern(C++) final class Semantic3Visitor : Visitor
                 {
                     /* Wrap the entire function body in a synchronized statement
                      */
-                    ClassDeclaration cd = funcdecl.isThis() ? funcdecl.isThis().isClassDeclaration() : funcdecl.parent.isClassDeclaration();
+                    ClassDeclaration cd = funcdecl.toParent4().isClassDeclaration();
```

Comment:
I should fix this. This doesn't work with static member template functions. I should probably make it a separate pull request since it was already erroneous before this change.

The following peace of code shouldn't fail.
```d
class C
{
    static synchronized func(alias a)() {}
}

void main()
{
    int a;
    C.func!a();
}
```